aips/aip-82.md

```diff
@@ -0,0 +1,149 @@
+
+## Alternative solutions
+
+* Require each provider to build their own specialized function and then have each provider build their own dispatch function. As mentioned above this quickly becomes untenable and not scalable.
+* Leverage dynamic dispatch of store creation. This is feasible, but we would rather receive feedback on this requirement rather than exposing more dynamic dispatch to the framework at this time. This would likely result in adhoc implementations that might not be secure.
+
+This solution unifies the solution and minimizes code.
+
+## Specification and Implementation Details
+
+This AIP introduces the following new functions:
+
+In `object.move`:
+```
+public fun root_owner<T: key>(object: Object<T>): address acquires ObjectCore {
+    let obj_owner = owner(object);
+    while (is_object(obj_owner)) {
+        obj_owner = owner(object::address_to_object<ObjectCore>(obj_owner));
+    };
+    obj_owner
+}
+```
+
+In `fungible_asset.move`:
+```
+/// Defnes a fungible asset stores as being untransferable at the object layer.
+#[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+struct Untransferable has key { }
+
+/// For untransferable stores allow deletion by the root owner.
+#[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+struct Deletable has key {
+    delete_ref: DeleteRef,
+}
+
+/// Set that only untransferable stores can be create for this fungible asset.
+public fun set_untransferable(constructor_ref: &ConstructorRef) {
+    let metadata_addr = object::address_from_constructor_ref(constructor_ref);
+    assert!(exists<Metadata>(metadata_addr), error::not_found(EFUNGIBLE_METADATA_EXISTENCE));
+    let metadata_signer = &object::generate_signer(constructor_ref);
+    move_to(metadata_signer, Untransferable {});
+}
+
+/// Creates an untransferable store.
+public fun create_untransferable_store<T>(account: &signer, metadata: Object<T>): Object<FungibleStore> {
+    let constructor_ref = object::create_object_from_account(account);
+    
+    // Freeze the object
+    let obj_transfer_ref = object::generate_transfer_ref(&constructor_ref);
+    object::disable_ungated_transfer(&obj_transfer_ref);
+    
+    if (object::can_generate_delete_ref(&constructor_ref)) {
+        let delete_ref = object::generate_delete_ref(&constructor_ref);
+        let store_signer = &object::generate_signer(&constructor_ref);
+        move_to(store_signer, Deletable { delete_ref });
+    };
+    create_store_internal(&constructor_ref, metadata)
+}
+
+/// Creates a store on an object during construction.
+public fun create_store<T: key>(
+    constructor_ref: &ConstructorRef,
```

Comment:
if we start with `ConstructorRef` instead of `&ConstructorRef`, this current solution would be much cleaner.

I propose to deprecate this function and use a new `create_fungible_store` function that takes `ConstructorRef` in a unified way and ban transfer based on whether `Untransferable` exists. The reason is for projects dealing with FA you don't want them to do 
```
if (it is transferrable) { create_store } else {create_untransferrable_store}... The api call should be transparent to users.

Reply:
Same issue, I can generate a `TransferRef` then pass in the `ConstructorRef` and bypass any logic here.